library.sh

```diff
@@ -429,45 +434,46 @@ function mktemp_with_extension() {
 function create_junit_xml() {
   local xml
   xml="$(mktemp_with_extension "${ARTIFACTS}"/junit_XXXXXXXX xml)"
-  echo "JUnit file ${xml} is created for reporting the test result"
+  echo "XML report for $1::$2 written to ${xml}"
   run_kntest junit --suite="$1" --name="$2" --err-msg="$3" --dest="${xml}" || return 1
 }
 
 # Runs a go test and generate a junit summary.
 # Parameters: $1... - parameters to go test
 function report_go_test() {
   local go_test_args=( "$@" )
-  # Install gotestsum if necessary.
-  run_go_tool gotest.tools/gotestsum gotestsum --help > /dev/null 2>&1
-  # Capture the test output to the report file.
-  local report
-  report="$(mktemp)"
-  local xml
+  local logfile xml ansilog htmllog
   xml="$(mktemp_with_extension "${ARTIFACTS}"/junit_XXXXXXXX xml)"
+  # Keep the suffix, so files are related.
+  logfile="${xml/junit_/go_test_}"
+  logfile="${logfile/.xml/.jsonl}"
   echo "Running go test with args: ${go_test_args[*]}"
-  capture_output "${report}" gotestsum --format "${GO_TEST_VERBOSITY:-testname}" \
-    --junitfile "${xml}" --junitfile-testsuite-name relative --junitfile-testcase-classname relative \
+  go_run gotest.tools/gotestsum@v1.8.0 \
+    --format "${GO_TEST_VERBOSITY:-testname}" \
+    --junitfile "${xml}" \
+    --junitfile-testsuite-name relative \
+    --junitfile-testcase-classname relative \
+    --jsonfile "${logfile}" \
     -- "${go_test_args[@]}"
-  local failed=$?
-  echo "Finished run, return code is ${failed}"
+  local gotest_retcode=$?
+  echo "Finished run, return code is ${gotest_retcode}"
 
   echo "XML report written to ${xml}"
-  if [[ -n "$(grep '<testsuites></testsuites>' "${xml}")" ]]; then
-    # XML report is empty, something's wrong; use the output as failure reason
-    create_junit_xml _go_tests "GoTests" "$(cat "${report}")"
-  fi
-  # Capture and report any race condition errors
-  local race_errors
-  race_errors="$(sed -n '/^WARNING: DATA RACE$/,/^==================$/p' "${report}")"
-  create_junit_xml _go_tests "DataRaceAnalysis" "${race_errors}"
-  if (( ! IS_PROW )); then
-    # Keep the suffix, so files are related.
-    local logfile=${xml/junit_/go_test_}
-    logfile=${logfile/.xml/.log}
-    cp "${report}" "${logfile}"
-    echo "Test log written to ${logfile}"
-  fi
-  return ${failed}
+  echo "Test log (JSONL) written to ${logfile}"
+
+  ansilog="${logfile/.jsonl/-ansi.log}"
+  go_run github.com/haveyoudebuggedit/gotestfmt/v2/cmd/gotestfmt@v2.3.1 \
+    -input "${logfile}" \
+    -showteststatus \
+    -nofail > "$ansilog"
+  echo "Test log (ANSI) written to ${ansilog}"
+
+  htmllog="${logfile/.jsonl/.html}"
+  go_run github.com/buildkite/terminal-to-html/v3/cmd/terminal-to-html@v3.6.1 \
+    --preview < "$ansilog" > "$htmllog"
+  echo "Test log (HTML) written to ${htmllog}"
```

Comment:
Rather than running these every time, how about providing instructions for users on how to do these if needed?

Reply:
The original JSON file isn't easy to understand for humans - see [example file](https://storage.googleapis.com/knative-prow/pr-logs/pull/knative_hack/172/unit-tests_hack_main/1519415863650816000/artifacts/go_test_xUZhYeE4.jsonl). Providing an easy-to-read HTML log file would greatly enhance the usability of test reports - see [example file](https://storage.googleapis.com/knative-prow/pr-logs/pull/knative_hack/172/unit-tests_hack_main/1519415863650816000/artifacts/go_test_xUZhYeE4.html)

I think most of the people will not know how to use the original JSON lines report file, and I don't see a place to post instructions on how to read it. The additional cost of creating the HTML report and keeping it on cloud storage isn't large. This is just a test output, kilobytes, even in the worst cases.